File: cmd/script/metrics.go
package main

import (
	"fmt"
	"net/http"
	"sync"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var (
	metricsRegistry = make(map[string]*AggregatorMetrics)
	metricsLock     sync.Mutex

	// Combined metric to track all aggregators in one place for easy comparison
	allAggregatorLatency *prometheus.GaugeVec

	// Pool discovery latency metric
	poolDiscoveryLatency *prometheus.GaugeVec
)

func init() {
	allAggregatorLatency = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "all_aggregator_latency_milliseconds",
			Help: "Latency in milliseconds for all aggregators by blockchain and source",
		},
		[]string{"aggregator", "chain"},
	)
	prometheus.MustRegister(allAggregatorLatency)

	poolDiscoveryLatency = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "pool_discovery_latency_milliseconds",
			Help: "Time from pool creation on-chain to first trade detection (pool discovery latency)",
		},
		[]string{"aggregator", "chain"},
	)
	prometheus.MustRegister(poolDiscoveryLatency)
}

type AggregatorMetrics struct {
	Latency *prometheus.GaugeVec
}

func GetOrCreateMetrics(aggregator string) *AggregatorMetrics {
	metricsLock.Lock()
	defer metricsLock.Unlock()

	if metrics, exists := metricsRegistry[aggregator]; exists {
		return metrics
	}

	metrics := &AggregatorMetrics{
		Latency: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: fmt.Sprintf("%s_latency_milliseconds", aggregator),
				Help: fmt.Sprintf("Latency in milliseconds for %s by blockchain", aggregator),
			},
			[]string{"chain"},
		),
	}

	prometheus.MustRegister(metrics.Latency)

	metricsRegistry[aggregator] = metrics
	return metrics
}

func RecordLatency(aggregator string, chain string, latencyMs float64) {
	metrics := GetOrCreateMetrics(aggregator)
	metrics.Latency.WithLabelValues(chain).Set(latencyMs)

	// Also record to the combined metric for easy comparison
	allAggregatorLatency.WithLabelValues(aggregator, chain).Set(latencyMs)
}

func RecordPoolDiscoveryLatency(aggregator string, chain string, latencyMs float64) {
	poolDiscoveryLatency.WithLabelValues(aggregator, chain).Set(latencyMs)
}

func StartMetricsServer(addr string) error {
	http.Handle("/metrics", promhttp.Handler())
	return http.ListenAndServe(addr, nil)
}

File: gmgn_monitor/metrics.py
"""
Prometheus metrics module for GMGN monitor
Similar to cmd/script/metrics.go
"""

from prometheus_client import start_http_server, Gauge


# Global metrics (similar to Go's package-level vars)
_swap_latency = None
_pool_discovery_latency = None
_all_aggregator_latency = None


def init_metrics():
    """Initialize Prometheus metrics (similar to Go's init())"""
    global _swap_latency, _pool_discovery_latency, _all_aggregator_latency

    _swap_latency = Gauge(
        'gmgn_latency_milliseconds',
        'Latency in milliseconds for GMGN swap trades by blockchain',
        ['chain']
    )

    _pool_discovery_latency = Gauge(
        'pool_discovery_latency_milliseconds',
        'Time from pool creation on-chain to first detection by GMGN',
        ['aggregator', 'chain']
    )

    _all_aggregator_latency = Gauge(
        'all_aggregator_latency_milliseconds',
        'Latency in milliseconds for all aggregators by blockchain and source',
        ['aggregator', 'chain']
    )


def record_latency(aggregator: str, chain: str, latency_ms: float):
    """
    Record latency metric
    Similar to Go's RecordLatency()

    Args:
        aggregator: Name of the aggregator (e.g., "gmgn")
        chain: Blockchain name (e.g., "solana", "base")
        latency_ms: Latency in milliseconds
    """
    # Filter out invalid values: negative or > 2 minutes (120000ms)
    if latency_ms < 0 or latency_ms > 120000:
        return

    # Record to aggregator-specific metric
    if _swap_latency:
        _swap_latency.labels(chain=chain).set(latency_ms)

    # Record to combined metric for easy comparison
    if _all_aggregator_latency:
        _all_aggregator_latency.labels(aggregator=aggregator, chain=chain).set(latency_ms)


def record_pool_discovery_latency(aggregator: str, chain: str, latency_ms: float):
    """
    Record pool discovery latency metric
    Similar to Go's RecordPoolDiscoveryLatency()

    Args:
        aggregator: Name of the aggregator (e.g., "gmgn")
        chain: Blockchain name (e.g., "solana", "base")
        latency_ms: Latency in milliseconds
    """
    # Filter out invalid values: negative or > 2 minutes (120000ms)
    if latency_ms < 0 or latency_ms > 120000:
        return

    if _pool_discovery_latency:
        _pool_discovery_latency.labels(aggregator=aggregator, chain=chain).set(latency_ms)

    # Also record to combined metric
    if _all_aggregator_latency:
        _all_aggregator_latency.labels(aggregator=aggregator, chain=chain).set(latency_ms)


def start_metrics_server(port: int):
    """
    Start Prometheus metrics HTTP server
    Similar to Go's StartMetricsServer()

    Args:
        port: Port number to listen on
    """
    start_http_server(port)
